core/field_variable.js

```diff
@@ -68,21 +77,30 @@ Blockly.FieldVariable.prototype.init = function() {
   this.initModel();
 };
 
+/**
+ * Initialize the model for this field if it has not already been initialized.
+ * If the value has not been set to a variable by the first render, we make up a
+ * variable rather than let the value be invalid.
+ * @package
+ */
 Blockly.FieldVariable.prototype.initModel = function() {
-  if (!this.getValue()) {
-    // Variables without names get uniquely named for this workspace.
-    var workspace =
-        this.sourceBlock_.isInFlyout ?
-            this.sourceBlock_.workspace.targetWorkspace :
-            this.sourceBlock_.workspace;
-    this.setValue(Blockly.Variables.generateUniqueName(workspace));
-  }
-  // If the selected variable doesn't exist yet, create it.
-  // For instance, some blocks in the toolbox have variable dropdowns filled
-  // in by default.
-  if (!this.sourceBlock_.isInFlyout) {
-    this.sourceBlock_.workspace.createVariable(this.getValue());
+  if (this.variable_) {
+    return; // Initialization already happened.
   }
+  this.workspace_ = this.sourceBlock_.workspace;
+  var variable = Blockly.Variables.getOrCreateVariable(
+      this.workspace_, null, this.defaultVariableName, this.defaultType_);
+  this.setValue(variable.getId());
+};
```

Comment:
I'm glad that variable creation is happening in initModel now, as this is directly applicable to the broadcast message use case in scratch-blocks. However, will it always be the case that the variable will be created on the source block's workspace? For example, in the scratch-blocks broadcast message use case, if the source block of the drop down is in the flyout, we want to create the new variable on the source block's target workspace instead. Is there a potential use case like this in blockly?

Reply:
Handled in `getOrCreateVariable`.  

" For example, in the scratch-blocks broadcast message use case, if the source block of the drop down is in the flyout, we want to create the new variable on the source block's target workspace instead."

Last time we chatted you said that the variable would not be created until the first time one of the referencing blocks was dragged from the flyout to the workspace.  This works in that case.  But until then it's a potential variable, not a real one.  Are you saying that's not the desired behaviour?

Reply:
When we chatted, I said that is the current state of things according to what I implemented. This is causing some weird behavior with creating new messages in the flyout (currently temporarily disabled in scratch-blocks master/develop): Creating a new message from the flyout (before dragging it out into the workspace) creates the new variable on the wrong workspace, and the menus of the blocks in the flyout don't get propagated correctly. The menu that was used to create the new message will only contain then new message and not message1 (since message1 hasn't been created as a variable yet) even though message1 existed in the drop down menu before the new message creation. 

I'm currently working on a lot of changes to broadcast blocks since I'm adding pluggable inputs to broadcast blocks (like in Scratch 2.0). I have mostly fixed this issue (and other related issues, some that came up with sb2 import of broadcast blocks with a plugged in input) and will have a PR out in the next day or two, so it's likely that this functionality does not need additional upstream blockly changes.

Reply:
Also, this is definitely not a required change in upstream blockly, as it is a very specific case in scratch-blocks. I only wanted to mention it in case there is a similar use case in blockly.

Reply:
I see.  Those failures make sense, and it sounds like you've already found the solutions:
- get the correct workspace [here](https://github.com/LLK/scratch-blocks/blob/develop/core/field_variable.js#L268)
- create `message1` when loading the workspace if you always want it to show up in the dropdown

I'll think about whether we're likely to have that case in Blockly.  My gut instinct is to say we won't, because it feels strange to have some variable types be created from the dropdown and other types from the flyout.  
On the other hand, creating a variable from the dropdown *should* be unambiguous about type as long as the dropdown only has one type in it.  And it's useful to be able to create and select the new variable at the same time from the block where you want to use it.  
This gets a lot harder when variable fields have multiple allowed types.